### PR TITLE
Search pattern generator + Visual compass head mode

### PR DESCRIPTION
**Pull request by Flova**
This pull request adds a generator for the head search pattern. Now the pattern can be modified on a higher level. This is pull request necessary for the Basler camera.
Additionally, the generator is used by the Visual Compass to generate a ground truth record pattern.

### CODE REVIEW

File: bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py [hunk 1]
```diff
@@ -33,19 +33,21 @@ def head_mode_callback(self, msg):
...
     #################
 
-    def send_motor_goals(self, pan_position, tilt_position, pan_speed=1.5, tilt_speed=1.5):
+    def send_motor_goals(self, pan_position, tilt_position, pan_speed=1.5, tilt_speed=1.5, clip=True):
         """
         :param pan_position: pan in radians
         :param tilt_position: tilt in radians
         :param pan_speed:
         :param tilt_speed:
+        :param no_clip:
```
Comment (timonegk) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py]: Parameter and documentation do not match

File: bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py [hunk 2]
```diff
@@ -63,3 +65,88 @@ def joint_state_callback(self, msg):
...
+            # Or go up/down
+            elif rightSide == rightDirection:
+                rightDirection = not rightDirection
+                if line in [0, lineCount - 1]:
+                    downDirection = not downDirection
+                if downDirection:
+                    line -= 1
+                else:
+                    line += 1
+        rospy.loginfo("Made pattern")
```
Comment (timonegk) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py]: Remove debug log

File: bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_kompass_record.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import math
+
+import rospy
+
+from std_msgs.msg import Header
+
+from dynamic_stack_decider.abstract_action_element import AbstractActionElement
+
+
+class VisualKompassRecord(AbstractActionElement):
```
Comment (timonegk) [inline bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_kompass_record.py]: Please rename Kompass to Compass

File: bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_kompass_record.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                                                                            interpolation_steps=self.interpolation_steps)
+
+        self.pattern_right = self.blackboard.head_capsule.generate_pattern(  self.blackboard.config['record_pattern_scan_lines'],
+                                                                            max(self.blackboard.config['record_pattern_pan_max_right']),
+                                                                            min(self.blackboard.config['record_pattern_pan_max_right']),
+                                                                            max(self.blackboard.config['record_pattern_tilt_max']),
+                                                                            min(self.blackboard.config['record_pattern_tilt_max']),
+                                                                            interpolation_steps=self.interpolation_steps)
+
+        self.visual_compass_trigger = rospy.Publisher( self.blackboard.config['vc_trigger_topic'], Header, queue_size=5)
```
Comment (timonegk) [inline bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_kompass_record.py]: This should be moved to `head_node.py` where the other publishers are registered, too.

File: bitbots_head_behavior/config/head_config.yaml [hunk 5]
```diff
@@ -7,16 +7,25 @@ behavior:
...
+    search_pattern_scan_lines: 4
+
+    # Values for the visual compass record pattern
+    record_pattern_scan_lines: 2
+    record_pattern_pan_max_left: [90, 20]
+    record_pattern_pan_max_right: [-20, -160]
+    record_pattern_tilt_max: [20, -10]
+
+    # Visual compass ground truth trigger topic
+    vc_trigger_topic: '/visual_compass_ground_truth_trigger'
```
Comment (timonegk) [inline bitbots_head_behavior/config/head_config.yaml]: Please rename to get rid of the 'vc' abbreviation.

File: bitbots_head_behavior/src/bitbots_head_behavior/actions/SearchPattern.py [hunk 6]
```diff
@@ -5,16 +5,23 @@
 from dynamic_stack_decider.abstract_action_element import AbstractActionElement
```
Comment (timonegk) [inline bitbots_head_behavior/src/bitbots_head_behavior/actions/SearchPattern.py]: Please keep the file names in snake case (search_pattern.py).

File: bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py [hunk 7]
```diff
@@ -63,3 +66,87 @@ def joint_state_callback(self, msg):
...
+    def _calculateHorizontalAngle(self, is_right, angle_right, angle_left):
+        """
+        The right/left position to an pan angle
+        """
+        if is_right:
+            return angle_right
+        else: 
+            return angle_left
+
+    def _interpolatedSteps(self, steps, tilt, min_pan, max_pan):
```
Comment (JanGut) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py]: Please add some documentation

File: bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py [hunk 8]
```diff
@@ -63,3 +66,87 @@ def joint_state_callback(self, msg):
...
+        delta = abs(min_pan - max_pan) 
+        step_size = delta / float(steps) 
+        output_points = list() 
+        for i in range(1, steps): 
+            value = int(i * step_size + min_pan) 
+            point = (value, tilt)     
+            output_points.append(point) 
+        return output_points 
+
+    def generate_pattern(self, lineCount, maxHorizontalAngleLeft, maxHorizontalAngleRight, maxVerticalAngleUp, maxVerticalAngleDown, interpolation_steps=0):
```
Comment (JanGut) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/head_capsule.py]: Please add some documentation

File: bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_compass_record.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                                                                            min(self.blackboard.config['record_pattern_pan_max_right']),
+                                                                            max(self.blackboard.config['record_pattern_tilt_max']),
+                                                                            min(self.blackboard.config['record_pattern_tilt_max']),
+                                                                            interpolation_steps=self.interpolation_steps)
+
+        # Append the two patterns and only take the first half of each
+        self.pattern = self.pattern_left + self.pattern_right
+        self.threshold = self.blackboard.config['position_reached_threshold']
+
+    def _notify_visual_compass(self):
```
Comment (JanGut) [inline bitbots_head_behavior/src/bitbots_head_behavior/actions/visual_compass_record.py]: Please add some documentation

**Pull request merged**